### PR TITLE
Documentation generation (via Sphinx)

### PR DESCRIPTION
**Pull request by KD-7**
# Implement Sphinx Doc Generation

## Necessary Steps:

- [x] Add sphinx config files
- [x] Resolve relative path errors 
- [x] Allow for automatic doc generation via github actions
- [x] Automatically push new documentation on release to github pages without the need to have a gh pages branch
- [x] Archive old versions of documentation
- [x] Reimplement deployment using read the docs to support versioned documentation
- [x] Make sphinx build fail on warning

## TODO in Future PR:

1. Fix doc build warnings
2. Remove requirements.txt dependency from build action

### Details
~~The new docs shall be deployed on github pages without the need to commit new docs to main on each successful PR merge.~~
Since we use read the docs, there is no need for gh pages

### CODE REVIEW

**Comment by yuxqiu** [comment]
@KD-7 I plan to convert our documentation from numpy style to google style as it is easier to write. By using google style, we don't need to maintain type information in the docstrings if we have type annotations. Also, the google style docstrings are more compact and easier to read. I would like to know if this breaks the current document generation.

**Comment by KD-7** [comment]
> @KD-7 I plan to convert our documentation from numpy style to google style as it is easier to write. By using google style, we don't need to maintain type information in the docstrings if we have type annotations. Also, the google style docstrings are more compact and easier to read. I would like to know if this breaks the current document generation.

No thats fine the sphinx extension napolean parses both numpy and google style docstrings. Whilst we are on that point there were some build errors: 
image
I think this is because doc strings are generated for the type hints (seperately) and then our user define docstrings which results in this warning.

Also, I will be changing the deployment/build to also make use of read the docs so we can show multiple versions of our documentation when we create a new release.

File: truelearn/__init__.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+"""major.minor.patch versioning convention
+==========================================
```
Comment (yuxqiu) [inline truelearn/__init__.py:2]: Replace this line with a blank line? Got a warning from `prospector`.

File: truelearn/__init__.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+major: breaking changes
+minor: new features
+patch: bug fixes
+"""
```
Comment (yuxqiu) [inline truelearn/__init__.py:6]: We could insert a blank line after this.

File: truelearn/__init__.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+__version__ = '1.1.1'
+
+__all__ = ['datasets', 'learning', 'models', 'preprocessing', 'tests', 'utils']
```
Comment (yuxqiu) [inline truelearn/__init__.py:9]: Let's stick with the double-quoted string as I've used it in the classifier and it makes more sense for someone with a C/C++ background.

File: truelearn/__init__.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+"""major.minor.patch versioning convention
```
Comment (yuxqiu) [inline truelearn/__init__.py:1]: Semantic Versioning? Perhaps we could just use "Semantic Versioning convention" here.

**Comment by KD-7** [comment]
@yuxqiu Documentation generation isnt fully finished: 
Due to the changes in #16 , it is starting to get very messy because I have made changes that rely on the pyproject.toml and its not found. Hence I think its best to merge this after #16 and then create a new branch to fix any additional errors.

Please could you review, thanks!

**Comment by yuxqiu** [comment]
> @yuxqiu Documentation generation isnt fully finished: Due to the changes in #16 , it is starting to get very messy because I have made changes that rely on the pyproject.toml and its not found. Hence I think its best to merge this after #16 and then create a new branch to fix any additional errors.
> 
> Please could you review, thanks!

Cool. I will do it now.

**Review by yuxqiu — COMMENTED** [review]
I think it's fine except for the same style issues. I think it would work better if you could add a space after `#`.

**Comment by KD-7** [comment]
> I think it's fine except for the same style issues. I think it would work better if you could add a space after `#`.

That should be all fixed now

**Review by yuxqiu — APPROVED** [review]

**Comment by yuxqiu** [comment]
> > @KD-7 I plan to convert our documentation from numpy style to google style as it is easier to write. By using google style, we don't need to maintain type information in the docstrings if we have type annotations. Also, the google style docstrings are more compact and easier to read. I would like to know if this breaks the current document generation.
> 
> No thats fine the sphinx extension napolean parses both numpy and google style docstrings. Whilst we are on that point there were some build errors: image I think this is because doc strings are generated for the type hints (seperately) and then our user defined docstrings which results in this warning. I haven't figured out yet the best course of action to resolve this.
> 
> Also, I will be changing the deployment/build to also make use of read the docs so we can show multiple versions of our documentation on our docs page for each truelearn release.

It seems that this duplicate object description problem is caused by the use of `dataclass`. Please see [How napoleon and autodoc interact documenting members](https://stackoverflow.com/questions/64588821/how-napoleon-and-autodoc-interact-documenting-members). This problem can be resolved by removing `:undoc-members:` from `truelearn.models.rst`.

**Pull request merged**